eclair-core/src/main/scala/fr/acinq/eclair/crypto/Sphinx.scala

```diff
@@ -74,298 +48,323 @@ object Sphinx extends Logging {
-    * @param sharedSecret shared secret for the sending node, which we will need to return error messages
+    * Peek at the first bytes of the per-hop payload to extract its length.
     */
-  case class ParsedPacket(payload: ByteVector, nextPacket: Packet, sharedSecret: ByteVector32)
-
-  /**
-    *
-    * @param privateKey     this node's private key
-    * @param associatedData associated data
-    * @param rawPacket      packet received by this node
-    * @return a ParsedPacket(payload, packet, shared secret) object where:
-    *         - payload is the per-hop payload for this node
-    *         - packet is the next packet, to be forwarded using the info that is given in payload (channel id for now)
-    *         - shared secret is the secret we share with the node that sent the packet. We need it to propagate failure
-    *         messages upstream.
-    */
-  def parsePacket(privateKey: PrivateKey, associatedData: ByteVector, rawPacket: ByteVector): Try[ParsedPacket] = Try {
-    require(rawPacket.length == PacketLength, s"onion packet length is ${rawPacket.length}, it should be ${PacketLength}")
-    val packet = Packet.read(rawPacket)
-    val sharedSecret = computeSharedSecret(PublicKey(packet.publicKey), privateKey)
-    val mu = generateKey("mu", sharedSecret)
-    val check = mac(mu, packet.routingInfo ++ associatedData)
-    require(check == packet.hmac, "invalid header mac")
-
-    val rho = generateKey("rho", sharedSecret)
-    val bin = (packet.routingInfo ++ ByteVector.fill(PayloadLength + MacLength)(0)) xor generateStream(rho, PayloadLength + MacLength + MaxHops * (PayloadLength + MacLength))
-    val payload = bin.take(PayloadLength)
-    val hmac = ByteVector32(bin.slice(PayloadLength, PayloadLength + MacLength))
-    val nextRouteInfo = bin.drop(PayloadLength + MacLength)
-
-    val nextPubKey = blind(PublicKey(packet.publicKey), computeblindingFactor(PublicKey(packet.publicKey), sharedSecret))
-
-    ParsedPacket(payload, Packet(Version, nextPubKey.value, hmac, nextRouteInfo), sharedSecret)
-  }
-
-  @tailrec
-  private def extractSharedSecrets(packet: ByteVector, privateKey: PrivateKey, associatedData: ByteVector32, acc: Seq[ByteVector32] = Nil): Try[Seq[ByteVector32]] = {
-    parsePacket(privateKey, associatedData, packet) match {
-      case Success(ParsedPacket(_, nextPacket, sharedSecret)) if nextPacket.isLastPacket => Success(acc :+ sharedSecret)
-      case Success(ParsedPacket(_, nextPacket, sharedSecret)) => extractSharedSecrets(nextPacket.serialize, privateKey, associatedData, acc :+ sharedSecret)
-      case Failure(t) => Failure(t)
+  def peekPayloadLength(payload: ByteVector): Int = {
+    payload.head match {
+      case 0 =>
+        // The 1.0 BOLT spec used 65-bytes frames inside the onion payload.
+        // The first byte of the frame (called `realm`) is set to 0x00, followed by 32 bytes of per-hop data, followed by a 32-bytes mac.
+        65
+      case _ =>
```

Comment:
Shouldn't there be a special case for `1` ?

Reply:
Why? 1 is an invalid value, but not at this layer, it will fail when trying to parse to a TLV stream.